### PR TITLE
Feat/crud movies

### PR DESCRIPTION
**Pull request by ahmedBassell**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -32,3 +32,4 @@ WORKDIR /home/node/app
 COPY --chown=node:node . .
 
 CMD [ "bash" ]
+CMD [ "npm", "start"]
```
Comment (MahaZaid) [inline Dockerfile:35]: add new line at the end of the file

File: docker-compose.yml [hunk 2]
```diff
@@ -6,8 +6,15 @@ services:
     container_name: emovis
     env_file:
       - ./.env
+    depends_on:
+      - mongo
```
Comment (MahaZaid) [inline docker-compose.yml:10]: we need to remove mongo I think & add mongoose image b2a

File: index.js [hunk 3]
```diff
@@ -1,4 +1,46 @@
 'use strict';
 
-// @TODO EXPORT main
-module.exports = {};
+const express = require('express')
+const bodyParser = require('body-parser');
+const connectDb = require("./src/connection");
+
+const app = express()
+const PORT = 3000
```
Comment (MahaZaid) [inline index.js:8]: let's read the PORT from `process.env` & add `PORT` to `env-example` file

File: index.js [hunk 4]
```diff
@@ -1,4 +1,46 @@
...
+
+// EMOJIS
+app.post('/emojis', (req, res) => emojis_controller.create(req, res))
+app.get('/emojis', (req, res) => emojis_controller.index(req, res))
+app.get('/emojis/:id', (req, res) => emojis_controller.show(req, res))
+app.put('/emojis/:id', (req, res) => emojis_controller.update(req, res))
+app.delete('/emojis/:id', (req, res) => emojis_controller.destroy(req, res))
+
+// RECOMMENDATION
+app.get('/recommend/:symbol', (req, res) => emovis_controller.recommend(req, res))
```
Comment (MahaZaid) [inline index.js:46]: add new line at the end

File: package.json [hunk 5]
```diff
@@ -23,12 +23,19 @@
...
     "docker": "docker-compose up -d --build && docker exec -it ${PWD##*/} bash",
     "test": "mocha ./test/"
   },
   "devDependencies": {
+    "nodemon": "^2.0.4"
   },
   "dependencies": {
+    "body-parser": "^1.19.0",
+    "express": "^4.17.1",
+    "mongodb": "^3.5.9",
```
Comment (MahaZaid) [inline package.json:38]: remove mongo from deps

File: src/Errors.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+class EmoviesError extends Error {
+
+    constructor(message) {
+        super(message);
+        this.name = 'EmoviesError';
+    }
+
+}
+
+class ValidationError extends EmoviesError {
```
Comment (MahaZaid) [inline src/Errors.js:12]: remove this class bardo

File: src/connection.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+const mongoose = require("mongoose");
+const Movie = require("./models/Movie.model");
+const connection = "mongodb://mongo:27017/emovis";
+const connectDb = () => {
+ return mongoose.connect(connection);
+};
+
+module.exports = connectDb;
```
Comment (MahaZaid) [inline src/connection.js:10]: add new line at the end

File: src/controllers/emojis.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+const Emoji = require("../models/Emoji.model");
+
+const EmojisController = {
+
+  async create(req, res) {
+    let symbol = req.body.symbol;
+    let genres = req.body.genres;
+
```
Comment (MahaZaid) [inline src/controllers/emojis.js:10]: I think we need to validate that `symbol ` & `genres ` are not undefined & they are one of the allowed values and if not throw a configuration error instance which is in `Errors` file

File: src/controllers/emojis.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  async show(req, res) {
+    let id = req.params.id;
+    const emoji = await Emoji.findById(id);
+
+    res.json(emoji);
+  },
+
+  async update(req, res) {
+    let id = req.params.id;
```
Comment (MahaZaid) [inline src/controllers/emojis.js:36]: we need to validate the `id` is not undefined

File: src/models/Movie.model.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+const movieSchema = new mongoose.Schema({
+ title: { type: String },
+ rating: { type: Number },
+ genres: { type: [String] },
+ description: { type: String },
+});
+
+const Movie = mongoose.model("Movie", movieSchema);
+console.log(Movie)
```
Comment (MahaZaid) [inline src/models/Movie.model.js:13]: remove console.log :D

File: src/tasks/import_movies.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
+const mongoose = require("mongoose");
```
Comment (MahaZaid) [inline src/tasks/import_movies.js:1]: add `use strict`

File: src/tasks/import_emojis.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
+const mongoose = require("mongoose");
```
Comment (MahaZaid) [inline src/tasks/import_emojis.js:1]: `use strict`

File: docker-compose.yml (hunk 2 above)
Comment (ahmedBassell) [reply docker-compose.yml:10]: that's the name of the image, not the npm dependency, right?

Comment (MahaZaid) [reply docker-compose.yml:10]: Ahh, docker image name

File: src/controllers/emojis.js [hunk 13]
```diff
@@ -1,39 +1,38 @@
...
-    }
-  },
-
-  async show(req, res) {
+    async create(req, res) {
+        let emoji = req.body.emoji;
+		let genres = req.body.genres;
+
+		if(!validateEmoji(emoji)) {
+			res.status(400);
```
Comment (ahmedBassell) [inline src/controllers/emojis.js:12]: should be 422, right?

File: index.js [hunk 14]
```diff
@@ -37,8 +37,8 @@ app.get('/genres', (req, res) => res.send('Hello genres!'))
 
 // EMOJIS
 app.post('/emojis', (req, res) => emojis_controller.create(req, res))
-app.get('/emojis', (req, res) => emojis_controller.index(req, res))
-app.get('/emojis/:id', (req, res) => emojis_controller.show(req, res))
+app.get('/emojis', (req, res) => emojis_controller.findAll(req, res))
```
Comment (ahmedBassell) [inline index.js:40]: index and show is rails style :smile:

**Pull request merged**